text/3634-scoped-impl-trait-for-type.md

```diff
@@ -0,0 +1,3436 @@
+}
+
+fn function2() -> impl Trait {
+    use impl Trait for () {}
+
+    {} // Binds on trailing `{}`-block used as expression.
+}
+```
+
+In this case, the returned opaque types use the respective inner scoped implementation, as it binds on the `()` expression.
+
+The following functions do not compile, as the implicitly returned `()` is not stated *inside* the scope where the implementation is available:
+
+```rust
+trait Trait {}
+
+fn function() -> impl Trait {
+                 ^^^^^^^^^^
+    use impl Trait for () {}
+    ---------------------
+
+    // Cannot bind on implicit `()` returned by function body without trailing *Expression*.
+}
+
+fn function2() -> impl Trait {
+                  ^^^^^^^^^^
+    use impl Trait for () {}
+    ---------------------
+
+    return; // Cannot bind on `return` without expression.
+    -------
+}
+```
+
+(The errors should ideally also point at the scoped implementations here with a secondary highlight, and suggest stating the return value explicitly.)
+
+The binding must be consistent:
+
+```rust
+trait Trait {}
+
+fn function() -> impl Trait {
+    // error: Inconsistent implementation of opaque return type.
+    if true {
+        use impl Trait for () {}
+        return ();
+        ----------
+    } else {
+        use impl Trait for () {}
+        return ();
+        ^^^^^^^^^^
+    }
+}
+```
+
+This function *does* compile, as the outer scoped `impl Trait for ()` is bound on the `if`-`else`-expression as a whole.
+
+```rust
+trait Trait {}
+
+fn function() -> impl Trait {
+    use impl Trait for () {}
+
+    if true {
+        use impl Trait for () {} // warning: unused scoped implementation
+        ()
+    } else {
+        use impl Trait for () {} // warning: unused scoped implementation
+        ()
+    }
+}
+```
+
+This compiles because the end of the function is not reachable:
+
+```rust
+trait Trait {}
+
+fn function() -> impl Trait {
+    {
+        use impl Trait for () {}
+        return (); // Explicit `return` is required to bind in the inner scope.
+    }
+}
+```
+
+## Static interception of dynamic calls
+
+As a consequence of binding outside of generic contexts, it *is* possible to statically wrap *specific* trait implementations on *concrete* types. This includes the inherent implementations on trait objects:
+
+```rust
+use std::fmt::{self, Display, Formatter};
+
+{
+    use impl Display for dyn Display {
+        fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+            // Restore binding to inherent global implementation within this function.
+            use ::{impl Display for dyn Display};
+
+            write!(f, "Hello! ")?;
+            d.fmt(f)?;
+            write!(f, " See you!")
+        }
+    }
+
+    let question = "What's up?"; // &str
+    println!("{question}"); // "What's up?"
+
+    let question: &dyn Display = &question;
+    println!("{question}"); // Binds to the scoped implementation; "Hello! What's up? See you!"
+}
+```
```

Comment:
This has to be explicitly special-cased to make it an error (also for supertraits, now that we have dyn-upcasting), but you're right. Even if someone later figures out a **really** good reason to have this, it's safer to start without it.

What about `&dyn Trait` types? Calling a function that takes `impl Trait` with a `&dyn Trait` seems like a common pattern for (mut-)ref-transitive traits like `Display` or `Write`, but this transitivity is never a built-in to my knowledge.

Reply:
Hm… this is actually pretty complicated.

Would this also exclude type aliases like `type Alias = dyn Trait;`? What about `type Alias = impl Trait;`?

That aside:

- Making it a hard error in all cases could interfere with blanket implementations like `use impl<T: ?Sized> Trait for T { … }` and equivalent imports.

- Making it silently ineffective for built-in trait implementations of trait objects would introduce subtle hole-punching:

  ```rust
  use a::Trait;
  use b::{impl<T: ?Sized> Trait for T};

  #[derive(Trait)] // Recursive-into-fields implementation.
  struct Struct {
      x: X,
      y: Y,
      z: dyn Trait,
  }
  ```

  Is the scoped implementation effective for recursion into `z` here? It would depend on implementation details of the derive-macro, as for example functions like

  ```rust
  pub Trait {
      fn by_ref<T: Trait>(reference: &T);
      fn by_value(value: impl Trait);
  }
  ```

  or alternatively functions like

  ```rust
  pub Trait {
      fn access(&self);
      fn consume(self) where Self: Sized;
  }
  ```

  would each bind differently between them when given a reference.

  (I am *assuming* that the derive-macro properly uses fully-qualified calls. If it uses method-style calls like `self.z.access()` and `self.z.consume()`, this might even differ by whether there's a `+ Copy` on the field type, though I'm not sure about that.)

Maybe the best course of action here would be to make it a **deny-by-default lint** that like other warnings is silenced if its span is resolved other-than-call-site? That would make everything behave consistently where there's an encapsulation boundary from an external macro, alert the user of the issue, but also provide an "I know what I am doing; this is intentional"-button in form of `#[allow(…)]`.

(Or at least _should_ behave like that for macros in an ideal world. In practice most proc macros are unhygienic in my experience, which is another case where a way to `#[allow(…)]` it may be needed.)

Reply:
The blanket impl `impl<T: ?Sized> Trait for T` already excludes `dyn Trait`, so there is no issue there. I would expect the equivalent `use impl` to also exclude `dyn Trait`. I would also prohibit scoped trait implementations of `Trait` on aliases for `dyn Trait`, and prohibit scoped implementations of `Trait` on type-alias `impl Trait`. I'm not as sure about scoped implementations of _other_ traits on type-alias `impl Trait`.

Reply:
Fixed in 99dc57251c6e03b8f333d1acf551fb6c0ab6936b.

(Thanks for explaining this to me.)